File: MATLAB_code/makelagmat.cpp
#include "slsc_mex.h"
#include "mex.h"
mxArray *makelagmat(mwSize numrows, mwSize numcols, int maxlag) {
    
    // Declare variables
    mxArray *elems, *lagmat, *lag, *lagdist, *tmp;
    mxLogical *elem;
    int32 *lm, i, j, k;
    float *ld;
    
    // Set up matrices
    elems  = mxCreateLogicalMatrix(numrows,numcols);
    lagmat = mxCreateNumericMatrix(numrows*numcols,numrows*numcols,mxINT32_CLASS,mxREAL);
    elem   = mxGetLogicals(elems);
    lm     = (int32 *)mxGetData(lagmat);
    lag    = mxCreateCellMatrix(maxlag,1);
    
    for (i = 0; i < numrows*numcols; i++) {
        // Turn element i on
        elem[i] = true;
        // Find all elements' distances from element i
        mexCallMATLAB(1, &lagdist, 1, &elems, "bwdist");
        ld = (float *)mxGetData(lagdist);
        
        // Fill in a column of data, only upper triangle
        for (j = 0; j <= i; j++)
            lm[j + i*mxGetM(lagmat)] = (int32)(round(ld[j]) - 1); // zero indexed
        
        // Switch element i off again
        elem[i] = false;
    }
    mxArray *lagptr;
    int32 *dataptr;
    int count,count1;
    for (k = 0; k < maxlag; k++) {
        // Figure out how large cell k of lag should be
        count = 0;
        for (i = 0; i < numrows*numcols; i++) {
            for (j = 0; j <= i; j++) {
                if (lm[j + i*mxGetM(lagmat)] == k)
                    count++;
            }
        }
        // Fill in dataptr with locations of lm == k
        lagptr  = mxCreateNumericMatrix(count, 1, mxINT32_CLASS, mxREAL);
        dataptr = (int32 *)mxGetData(lagptr);
        count1 = 0;
        for (i = 0; i < numrows*numcols; i++) {
            for (j = 0; j <= i; j++) {
                if (lm[j + i*mxGetM(lagmat)] == k)
                    dataptr[count1++] = j + i*mxGetM(lagmat) + 1;
                if (count1 == count)
                    break;
            }
            if (count1 == count)
                break;
        }
        // Set this data to cell k of lag
        mxSetCell(lag, k, lagptr);
    }
    
    return lag;
}

File: MATLAB_code/makelagmat.m
function lag = makelagmat(numrows,numcols,maxlag)
% MAKELAGMAT    Generates a cell array containing lag information
% A matrix that contains the lag number of each element with respect to
% every other element is output by this function. These values are
% stored in a NUMROWS*NUMCOLS by NUMROWS*NUMCOLS matrix. The output is a
% cell array containing the indices from the lag matrix that corresponds to
% the cell number, i.e. LAG{1} will contain all the indices of points in
% the lag matrix that are equal to 1, and LAG{2} for those that are equal
% to 2, etc.
%
% LAG = MAKELAGMAT(NUMROWS, NUMCOLS, MAXLAG) will return a cell array of
% dimensions (MAXLAG, 1). Each cell will contain the indices from the lag
% matrix that corresponds to the cell number.
% 
% LAG = MAKELAGMAT(NUMROWS, NUMCOLS) will assume that MAXLAG is set to
% the maximum possible value.

% Revision History
% 2011-05-09 dh65
%   Created function to be used with slsc_mex
if nargin == 2
    maxlag = round(sqrt(numrows^2+numcols^2))-1;
end

% Set up temp and lagmat
temp = false(numrows,numcols);
lagmat = zeros(numrows*numcols,numrows*numcols,'int32');

% Iterate through every element
for i = 1:numrows*numcols
    % "Turn on" the element so that bwdist can find all other elements'
    % distances from element i
    temp(i) = 1;
    lagdist = round(bwdist(temp));
    % Store information in lagmat
    lagmat(i,:) = lagdist(:);
    temp(i) = 0;
end
clear temp lagdist

% Transfer information from lagmat to lag
% Only save the upper triangle, since the matrix is symmetric
lag = cell(maxlag,1);
for i = 1:maxlag
    lag{i} = int32(find(triu(lagmat) == i));
end